Implement a "Self Directed Learning" mode to allow students to access unopened course contents

allenwq opened the pull request:
Fixes #2002 

![screen shot 2017-02-14 at 6 01 45 pm](https://cloud.githubusercontent.com/assets/4983239/22924356/be60e84c-f2df-11e6-9e86-5e250a8295f4.png)

- Add a course setting to allow enable and change the max time of the mode
- Redefine permissions to allow user to access unopened materials
- Update and further fix material permission. Before the material was accessible as long as the owner is started ( regardless of whether published or not )

weiqingtoh commented on app/controllers/course/controller.rb:
```diff
@@ -43,6 +43,11 @@ def current_component_host
 
+  # Override of Cancancan#current_ability to provide current course.
+  def current_ability
+    @current_ability ||= Ability.new(current_user, current_course)
```
Any particular reason why we need this?

weiqingtoh replied:
Oh ok, clearer when I check in the commits.

weiqingtoh commented on app/models/course/lesson_plan/item.rb:
```diff
@@ -40,6 +40,17 @@ def copy_attributes(other, time_shift)
+  # @return [Boolean]
+  def self_directed_started?
+    if course && course.self_directed_learning_max_time
+      start_at && start_at - course.self_directed_learning_max_time < Time.zone.now
```
Are there cases where `start_at` is not present?

allenwq replied:
Since we have database `null:false`, normally it won't. When you assign it to `nil` and call this method, it will break, here is just for safe.

weiqingtoh replied:
Ok 👍 once we move to Ruby 2.3 I think we can just use the safe navigation operator instead (`&.`)

fonglh replied:
we're already using Ruby 2.3.1.

weiqingtoh commented on app/models/course/lesson_plan/item.rb:
```diff
@@ -40,6 +40,17 @@ def copy_attributes(other, time_shift)
+  # Test if the lesson plan item has started for self directed learning.
+  #
+  # @return [Boolean]
+  def self_directed_started?
```
Would it be good practice to override `started?` instead, or should retain `self_directed_started?`

allenwq replied:
I was intended to do that also. but you know, each model has their own `started?` ( we have an extension for doing this),  so override `lesson_plan` won't work for assessment/video/survey/etc, as in they will still call their own versions.

weiqingtoh replied:
Yes on second thoughts I figured that it might unnecessarily override other models that `acts_as_lesson_plan_item`.

allenwq commented on app/models/course.rb:
```diff
@@ -113,6 +113,23 @@ def root_folder?
+    settings(:course).self_directed_learning_max_time
+  end
+
+  def self_directed_learning_max_time=(time)
```
@weiqingtoh Actually do you have better names for these methods?  now I feel the method names are too long and hard to recall in other places..

weiqingtoh replied:
Not sure, but something other than `self_directed_learning` would be good; it's too long. 

If we wanted to be more succinct, maybe `earlier_start_at` (or something like `earlier_start_interval`), so the code would be like this:

```ruby
if course && course.earlier_start_at
  start_at && start_at - course.earlier_start_at < Time.zone.now
else
```

Maybe can ask @fonglh also? 😅

fonglh replied:
How about `start_at_in_advance`? Or `advanced_start_at`?

allenwq replied:
@weiqingtoh @fonglh  this is the max `duration` of the mode.  so `acutal_start_at = start_at - duration`. but `earlier_start_at` or `advanced_start_at` are more like a time point ( other than a duration)

fonglh replied:
`advance_start_length`? or `advance_start_duration`?

allenwq replied:
changed to `advanced_start_at_duration`

allenwq commented on app/models/course.rb:
```diff
@@ -113,6 +113,28 @@ def root_folder?
+  def self_directed_learning_max_days=(value)
+    value = if value.present? && value.to_i > 0
+              value.to_i.days
+            else
```
Won't fix, I want it to be explicit that the value is set to `nil` in `else`

fonglh commented on config/locales/en/course/admin/admin.yml:
```diff
@@ -15,6 +15,12 @@ en:
+            Allow students to attempt the assessments that start at a future time provided they
+            have fulfilled the prerequisites.
+          self_directed_learning_time_html: >
+            Allow students to access course contents for %{input} days in advance.
```
"Allow students to access course items %{input} days in advance of the item's start time."

allenwq replied:
updated.

allenwq commented on app/models/course/lesson_plan/item.rb:
```diff
@@ -40,6 +40,17 @@ def copy_attributes(other, time_shift)
+  #
+  # @return [Boolean]
+  def self_directed_started?
+    if course&.advance_start_at_duration
```
sigh, hound doesn't recognise the new `&.` operator.

Going to update the config..

allenwq commented:
@fonglh @weiqingtoh Any more comments ?

Changes since last review:
- Changed the method name to `advance_start_at_duration` and some translations
- Upgraded to ruby 2.3 and `&.` operator is used.

fonglh commented on .travis.yml:
```diff
@@ -4,13 +4,13 @@ branches:
     - rvm: ruby-head
   exclude:
-    - rvm: 2.2.5
+    - rvm: 2.3.1
```
Think we should exclude the 2.3.3 one instead, because we're running 2.3.1.

allenwq replied:
OK 👍

weiqingtoh commented on client/app/bundles/course/admin/admin.js:
```diff
@@ -0,0 +1,13 @@
+$(document).ready(() => {
+  $('input.self-directed-learning-checkbox:checkbox').change(
```
Should we use `self-directed-learning` or `advance-start-at` here? 

I prefer the latter, as I'd think it is currently better for css to be consistent with backend column names (in the case where we need to do a global search and replace). 

When we move pages to react, that's when we can have different labelling for front-end elements and backend classes?

allenwq replied:
Will adopt `advance-start-at` to CSS classes, it's always good to keep them consistent 👍

weiqingtoh commented on app/models/components/course/materials_ability_component.rb:
```diff
@@ -63,4 +63,22 @@ def concrete_folder_hash
+  def opened_material_hashes
+    max_start_at = Time.zone.now
+    # Extend start_at time with self directed time from course settings.
+    max_start_at += (course.advance_start_at_duration || 0) if course
```
Would it be good to have this extracted as a helper function for all course components?

allenwq replied:
sorry, you mean the entire method or this line ?  don't really get you.

weiqingtoh replied:
Was trying to mean this line only - on second thoughts, my concern is not to have the `max_start_at` here, but have it reside as a method within the model file itself?

allenwq replied:
I think it's more clear to have it here and let the course only return a duration. 

You want the entire block to be in the `course.rb`? but for the courses which don't have a advance start_at, it's just `Time.zone.now`.

weiqingtoh replied:
I'm ok with leaving it here, I guess the issue is that the team needs to be mindful of this function (`self-directed-learning`) when working on future changes or new components.

allenwq replied:
that's why we must test the code and review each other's code :)

weiqingtoh replied:
I rather the code review process involve less changes, rather than rely only on the reviewers to catch missing pieces - ie. developers (new ones) could read the code and figure out some of these things on their own. 

Anyway, having this here is okay, so I think we can close this discussion.
